cmd/apply-crds/main.go

```diff
@@ -0,0 +1,163 @@
+	}
+}
+
+func main() {
+	ctx := context.Background()
+
+	initFlags()
+
+	config, err := getKubeConfig()
+	if err != nil {
+		log.Fatalf("Failed to get Kubernetes config: %v", err)
+	}
+
+	client, err := clientset.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("Failed to create API extensions client: %v", err)
+	}
+
+	// Walk the directory recursively and apply each YAML file.
+	err = filepath.Walk(crdsDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() || filepath.Ext(path) != ".yaml" {
+			return nil
+		}
+
+		log.Printf("Apply CRDs from file: %s", path)
+		if err := applyCRDsFromFile(ctx, client, path); err != nil {
+			log.Printf("Failed to apply CRD %s: %v", path, err)
+		}
+		return nil
+	})
+	if err != nil {
+		log.Fatalf("Error walking the path %s: %v", crdsDir, err)
+	}
+}
+
+// applyCRDsFromFile reads a YAML file, splits it into documents, and applies each CRD to the cluster.
+func applyCRDsFromFile(ctx context.Context, client *clientset.Clientset, filePath string) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("open file %q: %v", filePath, err)
+	}
+	defer file.Close()
+
+	// Create a decoder that reads multiple YAML documents.
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 4096)
+	var crdsToApply []*apiextensionsv1.CustomResourceDefinition
+	for {
+		crd := &apiextensionsv1.CustomResourceDefinition{}
+		if err := decoder.Decode(crd); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("decode YAML: %v", err)
+		}
+		crdsToApply = append(crdsToApply, crd)
+	}
+
+	// Apply each CRD separately.
+	for _, crd := range crdsToApply {
+		err := wait.ExponentialBackoffWithContext(ctx, retry.DefaultBackoff, func(context.Context) (bool, error) {
+			if err := applyCRD(ctx, client, crd); err != nil {
+				return false, nil
+			}
+			return true, nil
+		})
+		if err != nil {
+			return fmt.Errorf("apply CRD %s: %v", crd.Name, err)
+		}
+	}
+	return nil
+}
+
+// applyCRD creates or updates the CRD.
+func applyCRD(ctx context.Context, client *clientset.Clientset, crd *apiextensionsv1.CustomResourceDefinition) error {
+	crdClient := client.ApiextensionsV1().CustomResourceDefinitions()
+
+	// Check if CRD already exists in cluster and create if not found.
+	curCRD, err := crdClient.Get(ctx, crd.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		log.Printf("Create CRD %s", crd.Name)
+		_, err = crdClient.Create(ctx, crd, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("create CRD %s: %v", crd.Name, err)
+		}
+	} else {
+		log.Printf("Update CRD %s", crd.Name)
+		// Set resource version to update an existing CRD.
+		crd.SetResourceVersion(curCRD.GetResourceVersion())
+		_, err = crdClient.Update(ctx, crd, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("update CRD %s: %v", crd.Name, err)
+		}
+	}
+	return nil
+}
+
+func getKubeConfig() (*rest.Config, error) {
```

Comment:
I think we can replace that with https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client/config#GetConfig, unless we don't want to include controller runtime here

Reply:
I haven't used the controller-runtime dep on purpose. Usually we only need `rest.InClusterConfig()`, only for dev purposes we need the fallback.

Reply:
The `go.mod` has `sigs.k8s.io/controller-runtime v0.19.1` as a dep today, is this expected?

Reply:
Good call, the dep came after the implementation of the package. We have the controller-runtime dependency because of the envtest usage.
I think it's fair to use this instead then 👍🏻